modules/fedimint-wallet-server/src/lib.rs

```diff
@@ -1582,12 +1557,61 @@ mod tests {
 
     use bitcoin::Network::{Bitcoin, Testnet};
     use bitcoin::{Address, Amount, Network, OutPoint, Txid};
-    use fedimint_core::{BitcoinHash, Feerate};
-    use fedimint_wallet_common::{PegOut, PegOutFees, Rbf, WalletOutput};
+    use fedimint_core::{BitcoinHash, Feerate, PeerId};
+    use fedimint_wallet_common::{
+        PegOut, PegOutFees, Rbf, RoundConsensus, RoundConsensusItem, WalletOutput,
+    };
     use miniscript::descriptor::Wsh;
 
     use crate::common::PegInDescriptor;
-    use crate::{CompressedPublicKey, OsRng, SpendableUTXO, StatelessWallet, UTXOKey, WalletError};
+    use crate::{
+        CompressedPublicKey, OsRng, SpendableUTXO, StatelessWallet, UTXOKey, Wallet, WalletError,
+    };
+
+    fn round_item(block_height: u32, fee_rate: u64, random: u8) -> RoundConsensusItem {
+        RoundConsensusItem {
+            block_height,
+            fee_rate: Feerate {
+                sats_per_kvb: fee_rate,
+            },
+            randomness: [random; 32],
+        }
+    }
+
+    fn round_consensus(block_height: u32, fee_rate: u64, random: u8) -> RoundConsensus {
+        RoundConsensus {
+            block_height,
+            fee_rate: Feerate {
+                sats_per_kvb: fee_rate,
+            },
+            randomness_beacon: [random; 32],
+        }
+    }
+
+    #[test]
+    fn processes_round_consensus_items() {
+        // properly selects median fees / height and xor randomness
+        let consensus = Wallet::process_round_consensus(
+            0,
+            vec![
+                (PeerId::from(0), round_item(1, 4, 7)),
+                (PeerId::from(1), round_item(2, 5, 8)),
+                (PeerId::from(2), round_item(3, 6, 9)),
+            ],
+        );
+        assert_eq!(consensus, round_consensus(2, 5, 7 ^ 8 ^ 9));
+
+        // removes duplicate entries for a given peer, keeping last entry
+        let consensus = Wallet::process_round_consensus(
+            0,
+            vec![
+                (PeerId::from(0), round_item(1, 4, 7)),
+                (PeerId::from(1), round_item(2, 5, 8)),
+                (PeerId::from(1), round_item(3, 6, 9)),
+            ],
+        );
+        assert_eq!(consensus, round_consensus(3, 6, 7 ^ 9));
```

Comment:
For the block height and fees here, does the median calculation default to using a specific entry rather than averaging the numbers? (for instance, even when discounting Peer 1's first `RoundConsensusItem`, shouldn't the block height be an average of 1 and 3?)

Reply:
It's a median, so you throw out the duplicate, divide the number of items by 2 (`2/2=1`) so you get the last item.

Reply:
I think there are different definitions of medians, but this one was good enough when I thought it through.